SALTO-763 converted businessHours to MapType & resolve references to map values

nitzan-haim opened the pull request:
* converted BusinessHoursEntry values to a map (e.g Business_Hours_name (e.g 'Default'): { .. BusinessHoursEntry data ... }),
so that the field `businessHours` in `BusinessHoursSettings` will be of type `Map<BusinessHoursEntry>`.
* added resolving references of map values (to be able to resolve references of BusinessHoursEntrys)
adding the references upon fetch will be added in a separate PR. Also, adding a reference to a map value manually and then attempting to deploy causes a validation error, which will be fixed in a future PR.

netama commented on packages/salesforce-adapter/src/transformers/reference_mapping.ts:
```diff
@@ -82,6 +84,23 @@ const ReferenceSerializationStrategyLookup: Record<
+        const type = metadataType(ref.topLevelParent)
+        if (metadataTypeToFieldToMapDef[type] !== undefined
+           && field !== undefined
+           && metadataTypeToFieldToMapDef[type][field?.name] !== undefined) {
```
nit: can remove the `?` after the previous line's check

netama commented on packages/salesforce-adapter/src/transformers/reference_mapping.ts:
```diff
@@ -501,35 +528,42 @@ export const generateReferenceResolverFinder = (
       .map(def => def.serializationStrategy)
 
-    if (strategies.length > 1) {
+    if (strategies.length >= 1) {
```
nit: the goal of that debug log was just to point out that there are conflicts between the strategies that should be investigated, it's not very informative for the usual case of 1 strategy. might be better to keep it that way? (I think you could get the same return values with the old code)

netama commented on packages/salesforce-adapter/src/transformers/reference_mapping.ts:
```diff
@@ -82,6 +84,23 @@ const ReferenceSerializationStrategyLookup: Record<
+           && metadataTypeToFieldToMapDef[type][field?.name] !== undefined) {
+          const { key } = metadataTypeToFieldToMapDef[type][field.name]
+          // key is a field of ref.value which is also the key of ref.value in the map
+          return ref.value[key] ?? ref.value
```
is the fallback to `ref.value` a valid scenario? I'm not sure I understand when it happens. also, can you please add tests for this new strategy?

nitzan-haim replied:
it shouldn't happen, but I didn't want to leave it without a fallback. Do you have a better idea?

netama replied:
got it - in that case deploy would fail, right? I think it might be worth at least logging an error in that case, and not sure if it's better to fall back to a full object or to `undefined`. would either of them work? (I guess the same applies to the other `return ref.value` below)

netama commented on packages/salesforce-adapter/test/filters/profile_maps.test.ts:
```diff
@@ -15,7 +15,7 @@
 */
```
should probably also rename to `convert_maps`

netama reviewed (commented):
Looks good :) added a small question on the reference strategy that I didn't fully understand. Also:
1. This seems to be combining a few different issues (though they seem to be mixed in the ticket), wdyt about breaking this into separate PRs? (doesn't have to be too much - maybe one for extending the map filter and reference mechanism and another for the rest)
2. Please add tests for the new reference strategy. It might also be worth adding Entitlement Process to the e2e tests to make sure future changes that others make won't accidentally break deployments, but your call
3. Re the validation error - just making sure, does it only show up when a reference is added manually? it does not happen when running the new code on existing workspaces?

nitzan-haim commented on packages/salesforce-adapter/src/change_validators/profile_map_keys.ts:
```diff
@@ -70,7 +70,7 @@ const getMapKeyErrors = (
 
 const changeValidator: ChangeValidator = async changes => (
-  getProfileInstanceChanges(changes).flatMap(change => getMapKeyErrors(
+  getProfileInstanceChanges(changes, PROFILE_METADATA_TYPE).flatMap(change => getMapKeyErrors(
```
it's definitely what I meant to do (didn't notice that the auto-refactoring keeps the original function name) :)

nitzan-haim commented:
the following changes were moved to PR #1675 : 
* added missing references: entryStartDateField , SObjectType
* added some missing fields
* changed the path of workflow actions

netama commented on packages/salesforce-adapter/test/transformers/transformer.test.ts:
```diff
@@ -1587,6 +1588,35 @@ describe('transformer', () => {
+        expect(getLookUpName({
+          ref: new ReferenceExpression(
+            mockBusinessHoursInstance.elemID,
+            mockResolvedValue,
```
I might be missing something, but wouldn't the resolved value be the full instance (and any chance `mockBusinessHoursInstance` should have its `value` set to `mockResolvedValue`)?

netama replied:
or actually, sorry - I think maybe the referenced elem id should point to a nested part of the instance and not the full instance? (if that's the case, that raises a question of what should be the resolved value when we are referencing the full instance, because of this ambiguity)

nitzan-haim replied:
You're right, `ref.elemId` was not configured correctly in the test. I'm actually not using the elemId in the serialization function, so the test still passed. `ref.resolvedValue` should still point to the value in the map as written in the test. 
When we are referencing the full Element, `ref.elemId` would point to the that Element's elemID and the resolved value would point to the Element itself, as expected.

netama reviewed (approved):
I think there might be a small bug in ~the reference and~ the test ~(the reference should point to the instance and not its value, so `serialize` might need an extra nesting layer)~ (edit: it might just be the test). Please make sure that's working as intended + add more explicit handling for the fallback (which I think would only be caused by mis-configuration?) - other than that everything looks good, so approving

netama reviewed (approved):
🎈
